src/Microsoft.TemplateEngine.Core/ParameterSetBuilder.cs

```diff
@@ -0,0 +1,284 @@
+        {
+            var old = _resolvedValues[parameter];
+            _resolvedValues[parameter] = new EvalData(evaluatedParameterData);
+            if (old.InputDataState != InputDataState.Unset)
+            {
+                _resolvedValues[parameter].Value = old.Value;
+            }
+
+            _result = null;
+        }
+
+        public bool HasParameterValue(ITemplateParameter parameter) => _resolvedValues[parameter].InputDataState != InputDataState.Unset;
+
+        public void EvaluateConditionalParameters(ILogger logger)
+        {
+            List<EvalData> evaluatedParameters = _resolvedValues.Values.ToList();
+
+            var variableCollection = new VariableCollection(
+                null,
+                evaluatedParameters
+                    .Where(p => p.Value != null)
+                    .ToDictionary(p => p.ParameterDefinition.Name, p => p.Value));
+
+            EvalData[] variableCollectionIdxToParametersMap =
+                evaluatedParameters.Where(p => p.Value != null).Select(p => p).ToArray();
+
+            EvaluateEnablementConditions(evaluatedParameters, variableCollection, variableCollectionIdxToParametersMap, logger);
+            EvaluateRequirementCondition(evaluatedParameters, variableCollection, logger);
+        }
+
+        public IEvaluatedParameterSetData Build()
+        {
+            if (_result == null)
+            {
+                _result = new EvaluatedParameterSetData(
+                    this,
+                    _resolvedValues.Select(p => p.Value.ToParameterData()).ToList());
+            }
+
+            return _result!;
+        }
+
+        private void EvaluateEnablementConditions(
+            IReadOnlyList<EvalData> parameters,
+            VariableCollection variableCollection,
+            EvalData[] variableCollectionIdxToParametersMap,
+            ILogger logger)
+        {
+            Dictionary<EvalData, HashSet<EvalData>> parametersDependencies = new();
+
+            // First parameters traversal.
+            //   - evaluate all IsEnabledCondition - and get the dependecies between the parameters during doing so
+            foreach (EvalData parameter in parameters)
+            {
+                if (!string.IsNullOrEmpty(parameter.ParameterDefinition.Precedence.IsEnabledCondition))
+                {
+                    HashSet<int> referencedVariablesIndexes = new HashSet<int>();
+                    // Do not remove from the variable collection though - we want to capture all dependencies between parameters in the first traversal.
+                    // Those will be bulk removed before second traversal (traversing only the required dependencies).
+                    parameter.IsEnabledConditionResult =
+                        Cpp2StyleEvaluatorDefinition.EvaluateFromString(logger, parameter.ParameterDefinition.Precedence.IsEnabledCondition, variableCollection, referencedVariablesIndexes);
```

Comment:
imo, this should be `Generator.EvaluateFromString(logger, condition, parameterSet)`.
`Generator` defines the syntax of this condition so it picks up the kind of evaluator and builds variable collection etc.

Or we can make evaluating conditions a component that both Edge and Generator use.

Reply:
There is a new requirement to extract the used variables from the expression - so that we can build the dependency graph